File: dataportal-app/src/interfaces/Species.ts
import {SuccessApiResponse} from "./ApiResponse";

export interface Species {
    acronym: string;
    scientific_name: string;
    common_name: string;
    taxonomy_id: number;
}

export interface PaginatedResponse<T> {
    data: T[];
    total: number;
}

export type SpeciesApiResponse = SuccessApiResponse<Species>;
export type SpeciesListResponse = SuccessApiResponse<Species[]>;

File: dataportal_api/dataportal/models/species.py
"""
Species document model for Elasticsearch.

This module defines the SpeciesDocument class for indexing species information
in the METT Data Portal.
"""

from elasticsearch_dsl import Document, Text, Keyword, Integer, Boolean


class SpeciesDocument(Document):
    """Elasticsearch document for species information."""

    scientific_name = Text(fields={"keyword": Keyword()})
    common_name = Text()
    acronym = Keyword()
    taxonomy_id = Integer()
    enabled = Boolean()

    class Index:
        name = "species_index"

    def save(self, **kwargs):
        """Set `_id` as `acronym`."""
        self.meta.id = self.acronym
        return super().save(**kwargs)
